File: Languages/Go/bee_movie/bee_movie.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"unicode"
)

// Hello, World!
func hello() {
	fmt.Println("Hello, World!")
}

type LetterCount struct {
	Letter rune
	Count  int
}

func analyzeBeeMovie() {
	scriptPath := "../beemoviescript.txt"

	file, err := os.Open(scriptPath)
	if err != nil {
		fmt.Printf("File not found: %s\n", scriptPath)
		return
	}
	defer file.Close()

	fmt.Println("Bee Movie Script:")
	fmt.Println("--------------------------------------------------")

	letterCounts := make(map[rune]int)
	totalLetters := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		fmt.Println(line)

		// Count letters
		for _, char := range line {
			if unicode.IsLetter(char) {
				lower := unicode.ToLower(char)
				letterCounts[lower]++
				totalLetters++
			}
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Printf("Error reading file: %v\n", err)
		return
	}

	fmt.Println("--------------------------------------------------")
	fmt.Println("Analysis complete.")

	if totalLetters > 0 {
		// Convert map to slice and sort
		var counts []LetterCount
		for letter, count := range letterCounts {
			counts = append(counts, LetterCount{letter, count})
		}

		sort.Slice(counts, func(i, j int) bool {
			return counts[i].Count > counts[j].Count
		})

		fmt.Println("\nTop 3 most commonly used letters:")
		for i := 0; i < 3 && i < len(counts); i++ {
			fmt.Printf("%d. '%c': %d times\n", i+1, counts[i].Letter, counts[i].Count)
		}
	} else {
		fmt.Println("No letters found in the script.")
	}
}

func main() {
	hello()
	analyzeBeeMovie()
}

File: Languages/Rust/bee_movie.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::collections::HashMap;

// Hello, World!
fn hello() {
    println!("Hello, World!");
}

#[derive(Debug)]
struct LetterCount {
    letter: char,
    count: usize,
}

fn analyze_bee_movie() -> Result<(), Box<dyn std::error::Error>> {
    let script_path = "../beemoviescript.txt";
    let file = File::open(script_path)?;
    
    println!("Bee Movie Script:");
    println!("--------------------------------------------------");
    
    let mut letter_counts = HashMap::new();
    let mut total_letters = 0;
    
    // Read file line by line and print each line
    for line in io::BufReader::new(file).lines() {
        let line = line?;
        println!("{}", line);
        
        // Count letters
        for char in line.chars() {
            if char.is_alphabetic() {
                let lower = char.to_ascii_lowercase();
                *letter_counts.entry(lower).or_insert(0) += 1;
                total_letters += 1;
            }
        }
    }
    
    println!("--------------------------------------------------");
    println!("Analysis complete.");
    
    if total_letters > 0 {
        // Convert HashMap to vector and sort
        let mut counts: Vec<LetterCount> = letter_counts
            .into_iter()
            .map(|(letter, count)| LetterCount { letter, count })
            .collect();
        
        counts.sort_by(|a, b| b.count.cmp(&a.count));
        
        println!("\nTop 3 most commonly used letters:");
        for (i, lc) in counts.iter().take(3).enumerate() {
            println!("{}. '{}': {} times", i + 1, lc.letter, lc.count);
        }
    } else {
        println!("No letters found in the script.");
    }
    
    Ok(())
}

fn main() {
    hello();
    
    if let Err(e) = analyze_bee_movie() {
        eprintln!("Error: {}", e);
    }
}
